File: api-asn/csharp/src/api/asn/x509/ESubjectDirectoryAttributes.cs
using System;
using System.Reflection;
using System.Text;
using Com.Objsys.Asn1.Runtime;
using log4net;
using tr.gov.tubitak.uekae.esya.api.asn.cms;
using tr.gov.tubitak.uekae.esya.api.common.bundle;
using tr.gov.tubitak.uekae.esya.asn.util;
using tr.gov.tubitak.uekae.esya.asn.x509;
using Attribute = tr.gov.tubitak.uekae.esya.asn.x509.Attribute;
namespace tr.gov.tubitak.uekae.esya.api.asn.x509
{
    public class ESubjectDirectoryAttributes : BaseASNWrapper<SubjectDirectoryAttributes>
    {
        private static readonly ILog logger = LogManager.GetLogger(MethodBase.GetCurrentMethod().DeclaringType);

        public ESubjectDirectoryAttributes(SubjectDirectoryAttributes aObject)
            : base(aObject)
        {
        }

        public ESubjectDirectoryAttributes(EAttribute[] attributes)
            : base(new SubjectDirectoryAttributes())
        {
            mObject.elements = unwrapArray<Attribute, EAttribute>(attributes);
        }

        public override String ToString()
        {

            StringBuilder result = new StringBuilder();
            Attribute[] attributes = mObject.elements;
            if (attributes != null)
            {
                foreach (Attribute anAttrlar in attributes)
                {
                    result.Append(_attributeCikar(anAttrlar)).Append("\n");
                }
            }
            return result.ToString();
        }

        public Asn1OpenType[] getAttributeValue(Asn1ObjectIdentifier aOID)
        {
            Attribute[] attrAr = this.getObject().elements;
            for (int i = 0; i < attrAr.Length; i++)
                if (attrAr[i].type.Equals(aOID))
                    return attrAr[i].values.elements;
            return null;
        }

        public ERoleSyntax getRoleSyntax()
        {
            try
            {
                Asn1OpenType[] attributeValues = getAttributeValue(UtilSubjectDirectoryAttr.oid_at_role);
                if (attributeValues == null)
                    return null;
                return new ERoleSyntax(attributeValues[0].mValue);
            }
            catch (Exception ex)
            {
                logger.Error("Failed to Read RoleSyntax.", ex);
                return null;
            }
        }

        private String _attributeCikar(Attribute aAttribute)
        {
            Asn1ObjectIdentifier tip = aAttribute.type;
            Asn1OpenType[] degerler = aAttribute.values.elements;
            try
            {
                if (tip.Equals(UtilSubjectDirectoryAttr.oid_pda_countryOfCitizenship))
                {
                    return Resource.message(Resource.COUNTRYCITIZENSHIP) + " = " + _printableToString(degerler);
                }
                else if (tip.Equals(UtilSubjectDirectoryAttr.oid_pda_countryOfResidence))
                {
                    return Resource.message(Resource.COUNTRYRESIDENCE) + " = " + _printableToString(degerler);
                }
                else if (tip.Equals(UtilSubjectDirectoryAttr.oid_pda_dateOfBirth))
                {
                    return Resource.message(Resource.DATEOFBIRTH) + " = " + _dateToString(degerler);
                }
                else if (tip.Equals(UtilSubjectDirectoryAttr.oid_pda_gender))
                {
                    return Resource.message(Resource.GENDER) + " = " + _printableToString(degerler);
                }
                else if (tip.Equals(UtilSubjectDirectoryAttr.oid_pda_placeOfBirth))
                {
                    return Resource.message(Resource.PLACEOFBIRTH) + " = " + _directoryToString(degerler);
                }
                else if (tip.Equals(UtilSubjectDirectoryAttr.oid_at_role))
                {
                    return Resource.message(Resource.ROLE) + " = " + _roleToString(degerler);
                }/*else if(tip.equals(UtilSubjectDirectoryAttr.oid_win_upn))
       	 {
       		 
       	 }*/
            }
            catch (Exception aEx)
            {
                logger.Warn("Attribute degeri alınırken hata oluştu", aEx);
                return "";
            }

            return "";
        }

        private String _printableToString(Asn1OpenType[] aDegerler)
        {
            String attr = "";
            for (int i = 0; i < aDegerler.Length; i++)
            {
                Asn1OpenType value = aDegerler[i];
                Asn1DerEncodeBuffer encBuf = new Asn1DerEncodeBuffer();
                value.Encode(encBuf);
                Asn1PrintableString val = new Asn1PrintableString();
                Asn1DerDecodeBuffer decBuf = new Asn1DerDecodeBuffer(encBuf.MsgCopy);
                val.Decode(decBuf);
                attr += val.mValue;
            }
            return attr;
        }

        private String _dateToString(Asn1OpenType[] aDegerler)
        {
            String attr = "";
            for (int i = 0; i < aDegerler.Length; i++)
            {
                Asn1OpenType value = aDegerler[i];
                Asn1DerEncodeBuffer encBuf = new Asn1DerEncodeBuffer();
                value.Encode(encBuf);
                Asn1GeneralizedTime val = new Asn1GeneralizedTime();
                Asn1DerDecodeBuffer decBuf = new Asn1DerDecodeBuffer(encBuf.MsgCopy);
                val.Decode(decBuf);
                int year = val.Year;  //for Asn error
                attr += val.GetTime().ToLocalTime().ToString("d MMM yyyy");//val.ToString("d MMM yyyy");//new SimpleDateFormat("dd.MM.yyyy").format(val.getTime().getTime());
            }
            return attr;
        }

        private String _directoryToString(Asn1OpenType[] aDegerler)
        {
            String attr = "";
            for (int i = 0; i < aDegerler.Length; i++)
            {
                Asn1OpenType value = aDegerler[i];
                Asn1DerEncodeBuffer encBuf = new Asn1DerEncodeBuffer();
                value.Encode(encBuf);
                DirectoryString val = new DirectoryString();
                Asn1DerDecodeBuffer decBuf = new Asn1DerDecodeBuffer(encBuf.MsgCopy);
                val.Decode(decBuf);
                switch (val.ChoiceID)
                {
                    case DirectoryString._UTF8STRING:
                        {
                            attr += ((Asn1UTF8String)val.GetElement()).mValue;
                            break;
                        }
                    case DirectoryString._PRINTABLESTRING:
                        {
                            attr += ((Asn1PrintableString)val.GetElement()).mValue;
                            break;
                        }
                    case DirectoryString._TELETEXSTRING:
                        {
                            attr += ((Asn1T61String)val.GetElement()).mValue;
                            break;
                        }
                    case DirectoryString._UNIVERSALSTRING:
                        {
                            attr += ((Asn1UniversalString)val.GetElement()).mValue;
                            break;
                        }
                    case DirectoryString._BMPSTRING:
                        {
                            attr += ((Asn1BMPString)val.GetElement()).mValue;
                            break;
                        }
                    default:
                        break;
                }
            }
            return attr;
        }

        private String _roleToString(Asn1OpenType[] aDegerler)
        {
            String attr = "";
            for (int i = 0; i < aDegerler.Length; i++)
            {
                Asn1OpenType value = aDegerler[i];
                Asn1DerEncodeBuffer encBuf = new Asn1DerEncodeBuffer();
                value.Encode(encBuf);
                RoleSyntax val = new RoleSyntax();
                Asn1DerDecodeBuffer decBuf = new Asn1DerDecodeBuffer(encBuf.MsgCopy);
                val.Decode(decBuf);
                GeneralName rolName = val.roleName;
                if (rolName.ChoiceID == GeneralName._REGISTEREDID)
                {
                    int[] oid = ((Asn1ObjectIdentifier)rolName.GetElement()).mValue;
                    attr += "{";
                    for (int j = 0; j < oid.Length; j++)
                    {
                        attr += oid[j] + " ";
                    }
                    attr += "}";
                }

            }
            return attr;
        }

        public EExtension toExtension(bool aCritic)
        {
            return new EExtension(EExtensions.oid_ce_subjectDirectoryAttributes, aCritic, getBytes());
        }

    }
}

File: api-asn/src/tr/gov/tubitak/uekae/esya/api/asn/x509/ESubjectDirectoryAttributes.java
package tr.gov.tubitak.uekae.esya.api.asn.x509;

import com.objsys.asn1j.runtime.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import tr.gov.tubitak.uekae.esya.api.asn.BaseASNWrapper;
import tr.gov.tubitak.uekae.esya.api.asn.cms.EAttribute;
import tr.gov.tubitak.uekae.esya.api.common.ESYAException;
import tr.gov.tubitak.uekae.esya.api.common.bundle.cert.CertI18n;
import tr.gov.tubitak.uekae.esya.asn.util.UtilSubjectDirectoryAttr;
import tr.gov.tubitak.uekae.esya.asn.x509.*;

import java.io.IOException;
import java.text.SimpleDateFormat;

/**
 * @author ayetgin
 */
public class ESubjectDirectoryAttributes extends BaseASNWrapper<SubjectDirectoryAttributes> {
    private static final Logger logger = LoggerFactory.getLogger(ESubjectDirectoryAttributes.class);

    public ESubjectDirectoryAttributes(SubjectDirectoryAttributes aObject) {
        super(aObject);
    }

    public ESubjectDirectoryAttributes(EAttribute[] attributes) {
        super(new SubjectDirectoryAttributes());
        mObject.elements = BaseASNWrapper.unwrapArray(attributes );
    }

    @Override
    public String toString() {

        StringBuilder result = new StringBuilder();
        Attribute[] attributes = mObject.elements;
        if (attributes!=null){
            for (Attribute anAttrlar : attributes) {
                result.append(_attributeCikar(anAttrlar)).append("\n");
            }
        }
        return result.toString();
    }

    public Asn1OpenType[] getAttributeValue(Asn1ObjectIdentifier aOID) throws IOException, Asn1Exception {
        Attribute[] attrAr = this.getObject().elements;
        for (int i = 0; i < attrAr.length; i++)
            if (attrAr[i].type.equals(aOID))
                return attrAr[i].values.elements;
        return null;
    }

    public ERoleSyntax getRoleSyntax() {
        try {
            Asn1OpenType[] attributeValues = getAttributeValue(UtilSubjectDirectoryAttr.oid_at_role);
            if(attributeValues == null)
                return null;
            return new ERoleSyntax(attributeValues[0].value);
        } catch (Exception ex) {
            logger.error("Failed to Read RoleSyntax.", ex);
            return null;
        }
    }

    private String _attributeCikar(Attribute aAttribute) {
        Asn1ObjectIdentifier tip = aAttribute.type;
        Asn1OpenType[] degerler = aAttribute.values.elements;
        try {
            if (tip.equals(UtilSubjectDirectoryAttr.oid_pda_countryOfCitizenship)) {
                return CertI18n.message(CertI18n.COUNTRYCITIZENSHIP) + " = " + _printableToString(degerler);
            }
            else if (tip.equals(UtilSubjectDirectoryAttr.oid_pda_countryOfResidence)) {
                return CertI18n.message(CertI18n.COUNTRYRESIDENCE) + " = " + _printableToString(degerler);
            }
            else if (tip.equals(UtilSubjectDirectoryAttr.oid_pda_dateOfBirth)) {
                return CertI18n.message(CertI18n.DATEOFBIRTH) + " = " + _dateToString(degerler);
            }
            else if (tip.equals(UtilSubjectDirectoryAttr.oid_pda_gender)) {
                return CertI18n.message(CertI18n.GENDER) + " = " + _printableToString(degerler);
            }
            else if (tip.equals(UtilSubjectDirectoryAttr.oid_pda_placeOfBirth)) {
                return CertI18n.message(CertI18n.PLACEOFBIRTH) + " = " + _directoryToString(degerler);
            }
            else if (tip.equals(UtilSubjectDirectoryAttr.oid_at_role)) {
                return CertI18n.message(CertI18n.ROLE) + " = " + _roleToString(degerler);
            }/*else if(tip.equals(UtilSubjectDirectoryAttr.oid_win_upn))
       	 {
       		 
       	 }*/
        } catch (Exception aEx) {
            logger.warn("Attribute degeri alınırken hata oluştu", aEx);
            return "";
        }


        return "";
    }

    private String _printableToString(Asn1OpenType[] aDegerler) throws ESYAException {
        try {
            String attr = "";
            for (int i = 0; i < aDegerler.length; i++) {
                Asn1OpenType value = aDegerler[i];
                Asn1DerEncodeBuffer encBuf = new Asn1DerEncodeBuffer();
                value.encode(encBuf);
                Asn1PrintableString val = new Asn1PrintableString();
                Asn1DerDecodeBuffer decBuf = new Asn1DerDecodeBuffer(encBuf.getMsgCopy());
                val.decode(decBuf);
                attr += val.value;
            }
            return attr;
        } catch (Exception e) {
            throw new ESYAException(e);
        }
    }

    private String _dateToString(Asn1OpenType[] aDegerler) throws ESYAException {
        try {
            String attr = "";
            for (int i = 0; i < aDegerler.length; i++) {
                Asn1OpenType value = aDegerler[i];
                Asn1DerEncodeBuffer encBuf = new Asn1DerEncodeBuffer();
                value.encode(encBuf);
                Asn1GeneralizedTime val = new Asn1GeneralizedTime();
                Asn1DerDecodeBuffer decBuf = new Asn1DerDecodeBuffer(encBuf.getMsgCopy());
                val.decode(decBuf);

                attr += new SimpleDateFormat("dd.MM.yyyy").format(val.getTime().getTime());
            }
            return attr;
        } catch (Exception e) {
            throw new ESYAException(e);
        }
    }

    private String _directoryToString(Asn1OpenType[] aDegerler) throws ESYAException {
        try {
            String attr = "";
            for (int i = 0; i < aDegerler.length; i++) {
                Asn1OpenType value = aDegerler[i];
                Asn1DerEncodeBuffer encBuf = new Asn1DerEncodeBuffer();
                value.encode(encBuf);
                DirectoryString val = new DirectoryString();
                Asn1DerDecodeBuffer decBuf = new Asn1DerDecodeBuffer(encBuf.getMsgCopy());
                val.decode(decBuf);
                switch (val.getChoiceID()) {
                    case DirectoryString._UTF8STRING: {
                        attr += ((Asn1UTF8String) val.getElement()).value;
                        break;
                    }
                    case DirectoryString._PRINTABLESTRING: {
                        attr += ((Asn1PrintableString) val.getElement()).value;
                        break;
                    }
                    case DirectoryString._TELETEXSTRING: {
                        attr += ((Asn1T61String) val.getElement()).value;
                        break;
                    }
                    case DirectoryString._UNIVERSALSTRING: {
                        attr += ((Asn1UniversalString) val.getElement()).value;
                        break;
                    }
                    case DirectoryString._BMPSTRING: {
                        attr += ((Asn1BMPString) val.getElement()).value;
                        break;
                    }
                    default:
                        throw new ESYAException("Unknown ChoiceID");
                }
            }
            return attr;
        } catch (Exception e) {
            throw new ESYAException(e);
        }
    }

    private String _roleToString(Asn1OpenType[] aDegerler) throws ESYAException {
        try {
            String attr = "";
            for (int i = 0; i < aDegerler.length; i++) {
                Asn1OpenType value = aDegerler[i];
                Asn1DerEncodeBuffer encBuf = new Asn1DerEncodeBuffer();
                value.encode(encBuf);
                RoleSyntax val = new RoleSyntax();
                Asn1DerDecodeBuffer decBuf = new Asn1DerDecodeBuffer(encBuf.getMsgCopy());
                val.decode(decBuf);
                GeneralName rolName = val.roleName;
                if (rolName.getChoiceID() == GeneralName._REGISTEREDID) {
                    int[] oid = ((Asn1ObjectIdentifier) rolName.getElement()).value;
                    attr += "{";
                    for (int j = 0; j < oid.length; j++) {
                        attr += oid[j] + " ";
                    }
                    attr += "}";
                }

            }
            return attr;
        } catch (Exception e) {
            throw new ESYAException(e);
        }
    }

    public EExtension toExtension(boolean aCritic) throws ESYAException {
        return new EExtension(EExtensions.oid_ce_subjectDirectoryAttributes, aCritic, this);
    }

}
